Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,96 @@
+# Host startup hook
+
+For .NET Core 3+, we want to provide a low-level hook that allows
+injecting managed code to run before the main application's entry
+point. This hook will make it possible for the host to customize the
+behavior of managed applications during process launch, after they
+have been deployed.
+
+## Motivation
+
+This would allow hosting providers to define custom configuration and
+policy in managed code, including settings that potentially influence
+load behavior of the main entry point such as the
+`AssemblyLoadContext` behavior. The hook could be used to set up
+tracing or telemetry injection, to set up callbacks for handling
+Debug.Assert (if we make such an API available), or other
+environment-dependent behavior. The hook is separate from the entry
+point, so that user code doesn't need to be modified.
+
+## Proposed behavior
+
+The `DOTNET_STARTUP_HOOKS` environment variable can be used to specify
+a list of managed assemblies and type names that contain `public
+static void Initialize()` methods, which will be called in the order
+specified, before the `Main` entry point:
+
+```
+DOTNET_STARTUP_HOOKS=/path/to/StartupHook1.dll!StartupHookNamespace.StartupHookType1;/path/to/StartupHook2.dll!StartupHookNamespace.StartupHookType2
+```
+
+This variable is a list of absolute assembly paths, each with a type
+name following an exclamation mark. The list is delimited by the
+platform-specific path separator (`;` on Windows and `:` on Unix). It
+may not contain any empty entries or a trailing path separator.
+
+Setting this environment variable will cause each of the specified
+types' `public static void Initialize()` methods to be called in
+order, synchronously, before the main assembly is loaded. The
+environment variable will be inherited by child processes by
+default. It is up to the `StartupHook.dll`s and user code to decide
+what to do about this - `StartupHook.dll` may clear them to prevent
+this behavior globally, if desired.
+
+Specifically, hostpolicy starts up coreclr and sets up a new
+AppDomain. It then invokes a private method in
+`System.Private.CoreLib`, which will call each
+`StartupHookType.Initialize()` in turn synchronously. This gives
+`StartupHookType` a chance to set up new `AssemblyLoadContext`s, or
+register other callbacks. After all of the `Initialize()` methods
```

Comment:
Any limitations of this callback? Can new `Thread` be started? Will callback be able to get the startup into and path to app folder to obtain config files and make decision about what features to enable?

Reply:
@SergeyKanzhelev - The process will have already been started, so I don't see how the startup info could be useful at this point? Environment variables will be accessible and modifiable if the start up hooks don't want to propagate them to potential child processes.

Starting a new thread makes me nervous - I don't think there would be a limitation, but at that point it seems like the hook is doing more than it should. Do you have a scenario in mind?

Reply:
Startup info and environment variables is a way to configure the callback. Arguments from startup info may help to identify the process so proper configuration source will be used (you may configure for monitoring winform app ONLY started with specific parameters or such). Also it is typically to check for some settings from app folder. So ideally startup hook should know where to read appsettings file from.


Simplest example of task from the startup hook is sending notification to the remote endpoint that process got started. And/or getting some configuration from that remote location.

Example of a thread is to initiate performance counters collection thread or run dedicated thread to batch and send events to the server. Timers are not ideal for this purpose as timers can be affected by thread-pool starvation. So you will loose precision of counters or telemetry from the app that really has a problem.

Another example of thread is to start a implement a delayed initialization of monitoring. Let's say we want to use startup hook for Application Insights (AI). AI SDK can be used as a direct reference from customer code or via runtime injection. If it is used via runtime injection - we may choose to wait and see whether direct reference was taken by an application in `Main` method so we will not initialize any data collection from SDK enabled via runtime.

Reply:
To answer @SergeyKanzhelev  question, there are no limitiations on what the code can do.   Note however that the code DOES have to work with the runtime / framework  that was launched (bascialy it needs to be compatible).

Reply:
New threads can be started, but I agree with @AaronRobinsonMSFT that it would be best to avoid relying too much on the thread behavior. I updated the doc with some details on the threading behavior and caveats.

Reply:
> Also it is typically to check for some settings from app folder. So ideally startup hook should know where to read appsettings file from.

Does the hook have a good way to find out the name of the app being executed? I suspect that the hooks may be running too early right now before the app name is configured in CoreCLR.

Reply:
With the current design, there's no built-in way to get the app name, but it would be possible to get the app directory (`AppContext.BaseDirectory`, or `AppContext.GetData("APP_CONTEXT_BASE_DIRECTORY")`). This might be sufficient depending on the configuration mechanism used.

If getting the app name is important we could either:
- Set an additional app variable when initializing coreclr, before calling the startup hooks - maybe `APP_CONTEXT_APP_NAME`, and expose it as `AppContext.AppName`.
- Or, call the startup hooks on a later code path (somewhere in `coreclr_execute_assembly`, instead of from the host), where the app name is already known by the clr. We would then also have to decide how to expose it, likely something like `AppContext.AppName`.

Reply:
I think the first option I described above would be the way to go if we do make the app name available. The reasoning being that for other activation scenarios that don't go through `coreclr_execute_assembly`, like COM, we would still be able to support a startup hook on the current path. This leaves us the flexibility to expose the app name in a way that's independent of the startup hook feature for activation scenarios that do have an app.